Unified way for extensions to register metrics

jmartisk opened the pull request:
\+ metrics for Agroal

Work in progress, any feedback welcome, especially from extension maintainers who plan to introduce metrics...
@lhauspie

jmartisk committed:
Unified way for extensions to register metrics

loicmathieu reviewed (commented):
I have several questions about this:

- Metrics consumes resources on each call, so should we enable it by default? Enabling health by default is not an issue because it's only used when calls ...
- For your datasource implementation, I saw that  you develop one specific Metric class per metric. I have concern about this as it will not scale well when you want to expose 50 metrics for an extension.
- You didn't implements metrics by annotations like in my prototype, is it something that you plan to integrate next ?

lhauspie commented:
I'll try to use that on kafka-streams extension and give you feedback.
But I still have some work to add readiness check due to very interesting discussions on my #4793

lhauspie commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -219,18 +219,21 @@ quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
 quarkus.datasource.username=username-default
 quarkus.datasource.min-size=3
 quarkus.datasource.max-size=13
+quarkus.datasource.enable-metrics=true
```
I think this property should be `quarkus.datasource.metrics.enabled=true` to be consistent with health check e.g. `quarkus.kafka-streams.health.enabled=false`.

jmartisk replied:
This is the `enable-metrics` property that enables metric collection, not exposition, and it is per data source. It is already in Quarkus, I'm only adding it to the documentation example, because you need to enable metrics collection if you also enable metrics exposition and don't want to have all values equal to zero.

jmartisk commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -219,18 +219,21 @@ quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
 quarkus.datasource.username=username-default
 quarkus.datasource.min-size=3
 quarkus.datasource.max-size=13
+quarkus.datasource.enable-metrics=true
```
This is the `enable-metrics` property that enables metric collection, not exposition, and it is per data source. It is already in Quarkus, I'm only adding it to the documentation example, because you need to enable metrics collection if you also enable metrics exposition and don't want to have all values equal to zero.

jmartisk commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -307,4 +315,64 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+            /*
+             * TODO: what is the correct thing to do when metrics for the extension are enabled,
+             * but a data source has its metrics collection disabled? Should we log a warning that the values
+             * will be zero, or should we not publish metrics for this data source?
```
But on which data sources, if there are many? On all of them? And even if the user explicitly set `enable-metrics=false` on some data source? It would probably be fine to automatically enable it if the user didn't specify `enable-metrics` at all, but overriding the user's decision doesn't seem right.
 And I'm not sure if we have the facility to find out if the user specified `false` or didn't specify it at all (in which case it defaults to `false` too)

loicmathieu replied:
So we need to switch to a String boolean `enable-metrics` with a default of auto :

- true: always enabled
- false: never enabled
- auto: enabled if smallrye-metrics is in the classpath ..

Or we need to switch to an Optional config item, this should also do the trick

jmartisk replied:
With the optional config item, you mean that if it's undefined, we will treat is as "auto"? I think that would make sense. Sounds better to me than turning it into a string. 

Btw this could get affected by the incoming config changes https://github.com/quarkusio/quarkus/pull/5387, I'm particularly worried about the "Tighten rules about accessing run time configuration values from build steps.", but I'm not familiar with what exactly the changes will be

loicmathieu replied:
Yes, you're right, datasource `enable-metrics` is RUNTIME so cannot be used inside a build step.
Your warning must be emitter inside your Metric classes or inside a recorder that is used inside a RUN_TIME build step.

jmartisk replied:
Hm, I just changed that to inject the runtime config into the processor and only pass it to a recorder, but it seems https://github.com/quarkusio/quarkus/issues/5514 will forbid this usage. So I will change it again.

dmlloyd replied:
Injecting to the build step and passing it to the recorder is OK for now.  We don't yet support directly injecting the config into the recorder so that has to be implemented before we can take any other approach.

jmartisk commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -307,4 +315,64 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+
+    // TODO: introduce more metrics
+    @BuildStep
+    void registerMetrics(AgroalRuntimeConfig agroalRuntimeConfig,
```
I need to inject the runtime config, because the build config doesn't contain the `metrics-enabled` property which enables/disables metric collection, and we need to work with its value here.

jmartisk commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1238,6 +1238,123 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+2. The underlying library uses its own way for collecting metrics and makes them available at runtime using its own API.
+3. The underlying library does not provide metrics (or there is no library at all) and you need to insert some code in the extension's codebase that will collect the metrics.
+
+==== Case 1: The library uses MP Metrics
```
Hmmmmm, I'm thinking again whether that pattern is correct at all - will it really work?
See for example https://github.com/smallrye/smallrye-fault-tolerance/blob/2.1.2/implementation/src/main/java/io/smallrye/faulttolerance/metrics/BulkheadWaitRecorder.java#L55
I think a no-op registry returning null on this line will cause a NPE if we don't prevent this method from executing. It's generally quite a fragile pattern that can break due to a lot of things

loicmathieu replied:
@jmartisk so you think that the implementation of #5388 will not works ? If I understand your concern the Bulkhead functionality should generates a NPE.

Only fallback (inside the extension) and retry (inside the main integration test) are tested for fault-tolerance inside Quarkus. Maybe adding more test will give us confidence about it.

jmartisk replied:
Ok; after the discussion on that PR I added a few sentences about this to the end of "Case 1" section

loicmathieu replied:
Maybe adding a reference to `smallrye-fault-tolerane` extension will be good, as it's an extension author guide adding guidance as example implementation is a good thing.

jmartisk replied:
Ok, but I will add that after your PR is merged :)

jmartisk commented:
>     * Metrics consumes resources on each call, so should we enable it by default? Enabling health by default is not an issue because it's only used when calls ...

I'm not sure what exactly you mean by enabling it by default here. The default is that an application's pom.xml doesn't have a dependency on `quarkus-smallrye-metrics` so it will not get metrics. And when the dependency is present, I would assume that the user indeed wants metrics.

> 
>     * For your datasource implementation, I saw that  you develop one specific Metric class per metric. I have concern about this as it will not scale well when you want to expose 50 metrics for an extension.
That's a good point, we could probably move all the metric retrieval logic into one bigger application-scoped bean and then the `Callable`-s registered in the metric registry will be just anonymous instances that obtain a reference to that bean and call one of its methods (and pass the return value on). Would that sound better?

> 
>     * You didn't implements metrics by annotations like in my prototype, is it something that you plan to integrate next ?
We could do that but we first need to find out how to get some annotated metrics into the `vendor` metric registry without making the whole API a big mess. It will most likely need some changes in SmallRye Metrics side. So I thought we would start with the programmatic way.

jmartisk commented:
I just addressed @loicmathieu 's concern about the fact that each metric would need its own class - now it's just one class per metric type. See `AgroalGauge` and `AgroalCounter` classes. The processor passes the name of the metric that should be retrieved as a `String`. A lambda or something would be nicer, but we can't pass lambdas from the deployment code as they would not be available at runtime.

Also introduced the option to provide a custom metric object for all metric types (this is required for gauges and optional for other metric types). It's no longer a `Callable`, I realized we can use the MP Metrics interfaces directly, so the build item now expects a serializable implementation of `org.eclipse.microprofile.metrics.Gauge` etc.

Cleaned up the logic in `AgroalProcessor` to avoid too much code duplication.

Didn't yet address the mismatch between `metrics.enabled` globally and `enable-metrics` set per data source. I need to find out if we will indeed be able to read and especially change the runtime configuration after the incoming config rework

loicmathieu commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java:
```diff
@@ -0,0 +1,81 @@
+            if (dataSourceName == null) {
+                dataSource = CDI.current().select(AgroalDataSource.class).get();
+            } else {
+                dataSource = CDI.current().select(AgroalDataSource.class, new DataSourceLiteral(dataSourceName)).get();
```
Inside extensions will usually direclty use `Arc` instead of `CDI`.
Not sure it have a performance impact.

jmartisk replied:
Ok, changed that to Arc

loicmathieu commented on extensions/smallrye-metrics/spi/src/main/java/io/quarkus/smallrye/metrics/deployment/spi/MetricBuildItem.java:
```diff
@@ -0,0 +1,75 @@
+     * @param enabled Whether this metric is enabled
+     * @param tags The tags that will be applied to this metric
+     */
+    public MetricBuildItem(Metadata metadata, boolean enabled, Tag... tags) {
```
You should also add the `configRoot` field to your MetricBuildItem as it is done to Health.
The configRoot is not the same as the extension name (for agroal, the config root is datasource).

This is for a future usage to implements more complex enabling/disabling of extension metrics like 
```
quarkus.metrics.extensions=transaction,kafka
# enabled metrics for the transaction and kafka extension only
```

jmartisk replied:
Done

loicmathieu commented:
> A lambda or something would be nicer, but we can't pass lambdas from the deployment code as they would not be available at runtime.

Are you sure? I already seen some inside deployment code that pass callbacks at runtime like `io.quarkus.kafka.streams.deployment.KafkaStreamsHotReplacementSetup.OnMessage` but maybe it's not the same construct.

jmartisk commented:
> Are you sure? I already seen some inside deployment code that pass callbacks at runtime like `io.quarkus.kafka.streams.deployment.KafkaStreamsHotReplacementSetup.OnMessage` but maybe it's not the same construct.

I don't understand these technical details too much, but I'd say with `KafkaStreamsHotReplacementSetup` it works because that's used by devmode and the devmode-related code can use deployment classes. In my case I tried passing a `Function<String, Long>` implemented as a lambda, instead of just `String`, and I got a `NoClassDefFoundError` at runtime.

loicmathieu commented:
> I don't understand these technical details too much,  I'd say with `KafkaStreamsHotReplacementSetup` it works because that's used by devmode and the devmode-related code can use deployment classes.

You're right, I think you understand these technical details better than me :)

loicmathieu reviewed (changes requested):
As said, the ` enable-metrics` datasource config item is runtime so cannot be read inside a build step.

jmartisk commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1238,6 +1238,129 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+        metrics.produce(new MetricBuildItem(activeCountMetadata,
+                new ActiveCountGauge(dataSource.getName()),
+                agroalBuildTimeConfig.metricsEnabled,
+                new Tag("datasource", dataSource.getName())));
```
Good catch, was planning to add this but somehow forgot

emmanuelbernard commented:
> As said, the ` enable-metrics` datasource config item is runtime so cannot be read inside a build step.

Hum `quarkus.datasource.health.enabled` is build time, should we do the same?

jmartisk commented:
> > As said, the ` enable-metrics` datasource config item is runtime so cannot be read inside a build step.
> 
> Hum `quarkus.datasource.health.enabled` is build time, should we do the same?

Do you mean turn `quarkus.datasource.DS_NAME.enable-metrics` into a build time property? (It's the one that enables metric collection per each data source)
Are you sure you're not confusing it with `quarkus.datasource.metrics.enabled`, which *is* a build time property and enables metric exposition in MP Metrics for the whole extension?
Damn, these two properties are really similar, a lot of people will be confused. We need to make sure we have very clear docs.
I would rather name the `metrics.enabled` property `metrics.exposed`, but then it would not be consistent with what other extensions have (unless the other extensions do it too).

emmanuelbernard commented:
> > > As said, the ` enable-metrics` datasource config item is runtime so cannot be read inside a build step.
> > 
> > 
> > Hum `quarkus.datasource.health.enabled` is build time, should we do the same?
> 
> Do you mean turn `quarkus.datasource.DS_NAME.enable-metrics` into a build time property? (It's the one that enables metric collection per each data source)
> Are you sure you're not confusing it with `quarkus.datasource.metrics.enabled`, which _is_ a build time property and enables metric exposition in MP Metrics for the whole extension?
> Damn, these two properties are really similar, a lot of people will be confused. We need to make sure we have very clear docs.
> I would rather name the `metrics.enabled` property `metrics.exposed`, but then it would not be consistent with what other extensions have (unless the other extensions do it too).
> Or we could rename `enable-metrics` to `collect-metrics` or something, but that would be a breaking change and I'm not sure if it would help much

All of the properties with `health` in them according to https://quarkus.io/guides/all-config are build time.
So I'm wondering why enable-metrics should be a runtime property?

Note that we can't rename properties now, we are in the 1.0+ phase.

jmartisk commented:
Perhaps it was made a runtime property because Agroal supports enabling/disabling metrics dynamically: https://github.com/agroal/agroal/blob/1.7/agroal-api/src/main/java/io/agroal/api/configuration/AgroalDataSourceConfiguration.java#L33

IIUC Quarkus is not making use of this at the moment, it treats the data source configuration as immutable once the data source is built. But we could detect a change in the config value and dynamically update the data source's configuration without having to create a new instance.

emmanuelbernard commented:
> Perhaps it was made a runtime property because Agroal supports enabling/disabling metrics dynamically: https://github.com/agroal/agroal/blob/1.7/agroal-api/src/main/java/io/agroal/api/configuration/AgroalDataSourceConfiguration.java#L33
> 
> IIUC Quarkus is not making use of this at the moment, it treats the data source configuration as immutable once the data source is built. But we could detect a change in the config value and dynamically update the data source's configuration without having to create a new instance.

I don't think we need such a feature. It was most likely an oversight.

jmartisk commented:
@emmanuelbernard  Yep, I agree. I'll change it to a build time property and attach it to this PR

jmartisk committed:
Turn datasources' metrics-enabled into a build time property

jmartisk committed:
Expose metrics for Agroal extension

gunnarmorling commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1238,6 +1238,130 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+==== Case 2: The library provides its own metric API
+In this case, the extension can make use of the `quarkus-smallrye-metrics-spi` module to expose the metrics from the library using MP Metrics.
+
+The extension author should to the following:
```
```suggestion
The extension author should do the following:
```

jmartisk replied:
fixed in my commit

gunnarmorling commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1238,6 +1238,130 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+
+[source%nowrap.java]
+----
+@BuildStep
```
@lhauspie, what's your take on this? IIUC, it'd require the KafkaStreams extension to list all the metrics explicitly, instead of the generic implementation we currently have (in the blog post). It seems that'd be quite a bit more work. WDYT?

gunnarmorling commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1238,6 +1238,130 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+- Metrics are not disabled for this extension
+
+A recommended way to implement this check would be to produce a specific `SystemPropertyBuildItem` during build, where it is easy
+to evaluate these two conditions, and then check back to this system property at runtime.
```
Perhaps show an implementation of `metricsEnabled`?

jmartisk replied:
I added an example how to produce the system property

gunnarmorling commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private AgroalDataSource getDataSource() {
+        if (dataSource == null) {
```
This double-check locking idiom makes me think `dataSource` should be volatile and also be read into a local variable.

loicmathieu replied:
+1 `dataSource` must be volatile here. 
But I will avoid using synchronization here and go for getting the datasource from an `@PostConstruct` block

jmartisk replied:
I guess you're right. I'm not sure if there's any better way to do this, I would need a callback that gets invoked after the Arc container is started, but before any metric requests arrive. I don't want to turn this into a CDI bean just for the purpose of being able to do `@PostConstruct`, due to the performance penalty, and also potentially breaking user applications by unexpectedly introducing some CDI beans that implement `Gauge` or `Counter`. I will fix the synchronization for now, perhaps we will find a better solution where this is not needed.

gunnarmorling reviewed (commented):
Some comments inline.

jmartisk commented:
Summary of the recent changes that I just did:
- `metrics-enabled` (enables collection of metrics internally in agroal) is now a build time property and it is now an `Optional`. If unspecified, it will be treated as `true` only if MP Metrics are enabled (which are enabled by default if the `smallrye-metrics` extension is present). So I also removed the WARN log that was in place for the cases where MP Metrics are enabled, but Agroal metrics are disabled for some data sources.
- Fixed synchronization in Agroal metric classes per @gunnarmorling 's comment
- Updated the docs with an example how to produce a system property to be used by frameworks to detect whether they should collect metrics. This references `Capabilities.METRICS` capability so this PR is now basically dependent on https://github.com/quarkusio/quarkus/pull/4987

loicmathieu commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private AgroalDataSource getDataSource() {
+        AgroalDataSource dsLocal = dataSource;
```
It's ot mandatory to store the attribute inside a local variable. Volatile is enought to make it correct.
See the last example on https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

jmartisk replied:
I got this from the interview with Joshua Bloch https://www.oracle.com/technical-resources/articles/javase/bloch-effective-08-qa.html, also the book Effective Java says that thanks to the local variable you get about a performance gain, because you only need to access the volatile field once if it's already non-null.  From wikipedia:
> Note the local variable "localRef", which seems unnecessary. The effect of this is that in cases where helper is already initialized (i.e., most of the time), the volatile field is only accessed once (due to "return localRef;" instead of "return helper;"), which can improve the method's overall performance by as much as 25 percent.

loicmathieu replied:
I never thought about these 2 volatile reads, I really need to read again Effective Java ;)
Thanks for the pointers !

loicmathieu commented:
@jmartisk moving the property from runtime to build time is a breaking change so I added the noteworthy label. Anyway adding new metrics worth a note :)

The only concern that I still have is the fact that the metrics are enabled by default. Computing metrics have a cost that is way more than health check. So even if the user enables metrics I don't think each extension should enable it's metrics by default.
I can totaly think of enabling metrics inside my application but didn't care of metrics from datasource or Kafka.

jmartisk commented:
@loicmathieu  So your concern is that if a user adds the extension, he might not want to have metrics for all extensions, but just some selected ones for which they will enable them? I don't know, that could be a valid point.
What do others think about this?

There's currently an issue that the agroal internal metrics get enabled automatically even if `smallrye-metrics` extension is not there, that's why the CI failed. I am planning to fix that. The problem is that to determine whether they should be enabled, it's no longer enough to just know the value of `quarkus.datasource.DS_NAME.enable-metrics` property, you also need to know whether the metrics capability is present... But if I add a check for this, this PR will become fully dependent on https://github.com/quarkusio/quarkus/pull/4987  so I would like if we could speed up merging that one :)

jmartisk commented:
- Fixed the issue with metrics collection getting enabled even if `quarkus-smallrye-metrics` is not present. This however means this PR will not compile until merging https://github.com/quarkusio/quarkus/pull/4987 and rebasing
- Added some text about metrics into Agroal docs

jmartisk commented:
Applied a change so that metrics from extensions should be disabled by default.
Rebased on top of the PR on which this depends so it will compile now.
Squashed commits into some reasonable chunks.
I think this should be ready to evolve from draft

loicmathieu reviewed (approved):
As metrics are disabled by default for extension, we need a way to tell the user that it exist.

I'm thinking of two things that we can do:

- Explain on the metric page that some extension may provides metrics and that this needs to be enabled (like we did for health but saying it's disabled by default: https://quarkus.io/guides/microprofile-health#extension-health-checks)
- Add a log in INFO in an extension providing metrics when smallrye-metrics is included but the extension metrics are disabled.

Anyway, it's all good to me, good job ;)

emmanuelbernard commented:
I think the metrics documentation should be in each extension providing these metrics, that's where you want to see them I suppose.

jmartisk commented:
Yeah I think each extension that supports it can have a short paragraph about it in its documentation. I included exactly that in the agroal documentation.

loicmathieu commented:
@emmanuelbernard @jmartisk keep in minds that developers often don't read documentation untill the end ... I don't :) Usually I read the first section then start coding and go back to documentation if I cannot make something works by myself. So some (a lot) may miss the section on the extension documentation.

It's important to also have a section, early in the metrics guide (maybe before the example application), explaining that a lot of Quarkus extension provides metrics that are disabled by default and the before implementing your own metrics you should check the extension documentation ...

emmanuelbernard commented:
Sure can double reference.

On Tue 26 Nov 2019 at 09:47, Loïc Mathieu <notifications@github.com> wrote:

> @emmanuelbernard <https://github.com/emmanuelbernard> @jmartisk
> <https://github.com/jmartisk> keep in minds that developers often don't
> read documentation untill the end ... I don't :) Usually I read the first
> section then start coding and go back to documentation if I cannot make
> something works by myself. So some (a lot) may miss the section on the
> extension documentation.
>
> It's important to also have a section, early in the metrics guide (maybe
> before the example application), explaining that a lot of Quarkus extension
> provides metrics that are disabled by default and the before implementing
> your own metrics you should check the extension documentation ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/5272?email_source=notifications&email_token=AACJNWFTE4QQAUTYKZXKVSTQVTPB5A5CNFSM4JKCKP7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFFG4KA#issuecomment-558525992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACJNWGQ2GDMNJUWUJ63K53QVTPB5ANCNFSM4JKCKP7A>
> .
>

jmartisk commented:
Ok, added a short paragraph to the beginning of the general metrics guide

jmartisk commented:
Hmm, CI started failing due to dependency divergence. I'll release SmallRye Metrics 2.3.2 to align the MP Metrics API dependency correctly. Should be ready within a few hours.

jmartisk commented:
After upgrading to SR-M 2.3.2, I faced another divergence problem, this time for Fault Tolerance extension depending on MP Metrics 2.2 (while the rest is at 2.2.1). I employed a not very nice solution to this by overriding the dependency in the Fault Tolerance extension to keep the enforcer plugin happy. I also raised a thread in SmallRye group (https://groups.google.com/forum/#!topic/smallrye/TitSifqEXvU) to find a better long term strategy to avoid having to do these things.

jmartisk commented:
I was hoping that after my paternity leave, this would be merged :) anyone have any more comments on it?

loicmathieu commented:
@jmartisk I already approve the PR so all is OK for me

jmartisk commented:
I just removed the workaround mentioned in my comment  because we now have updated SR Fault Tolerance that aligns correctly.
Would be nice to have this acked+merged soon because the work to get metrics working in other extensions depends on this, I'd like to start working on them. @emmanuelbernard does this have your ack?

emmanuelbernard commented:
Sorry guys, I lost track a bit so don't wait for me. I'll remove me from the equation by acking the review

emmanuelbernard commented:
@jmartisk but... are you talking about the fault tolerance that no longer use Hystrix that @michalszynkiewicz mentioned? Because that one will  land in 1.2. so that PR will do too.

jmartisk commented:
@emmanuelbernard No, in this case this was just a micro update to SR-FaultTolerance 2.1.3, this still depends on Hystrix

gunnarmorling commented:
Hey there @loicmathieu, @lhauspie, has either of you already looked into (or is planning to) using this for Kafka Streams?

jmartisk commented:
I already have metrics for Hibernate ORM almost ready, they depend on this mechanism too, so I would really appreciate if we could merge this soon.
CI failures are unrelated

jmartisk commented:
Rebased and fixed a conflict. @lhauspie have you had a chance to look please? I'm growing a bit impatient here :)

gsmet commented:
@jmartisk I'll have a look today.

gsmet commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -265,6 +265,24 @@ If you have multiple datasources, all datasources will be checked and the status
+expose metrics if the mechanism to collect them is disabled.
+
+Conversely, setting `quarkus.datasource.DATASOURCE_NAME.enable-metrics` to `true` explicitly can be used to enable collecting of metrics even if
+`quarkus-smallrye-metrics` extension is not in use. This can be useful if you need to access the collected metrics programmatically.
```
```suggestion
the `quarkus-smallrye-metrics` extension is not in use. This can be useful if you need to access the collected metrics programmatically.
```

gsmet commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -265,6 +265,24 @@ If you have multiple datasources, all datasources will be checked and the status
+For the exposed metrics to contain any actual values, it is necessary that metric collection is enabled internally
+by Agroal mechanisms. By default, this metric collection mechanism gets turned on for all data sources if the `quarkus-smallrye-metrics`
+is present and metrics for the Agroal extension are enabled. If you want to disable metrics for a particular data source,
+this can be done by setting `quarkus.datasource.DATASOURCE_NAME.enable-metrics` to `false`. This disables
```
```suggestion
this can be done by setting `quarkus.datasource.enable-metrics` to `false` (or `quarkus.datasource.<datasource name>.enable-metrics` for a named datasource). This disables
```

gsmet commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -265,6 +265,24 @@ If you have multiple datasources, all datasources will be checked and the status
+collecting the metrics as well as exposing them in the `/metrics` endpoint, because it does not make sense to
+expose metrics if the mechanism to collect them is disabled.
+
+Conversely, setting `quarkus.datasource.DATASOURCE_NAME.enable-metrics` to `true` explicitly can be used to enable collecting of metrics even if
```
```suggestion
Conversely, setting `quarkus.datasource.enable-metrics` to `true` (or `quarkus.datasource.<datasource name>.enable-metrics` for a named datasource) explicitly can be used to enable collecting of metrics even if
```

gsmet commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1290,6 +1290,143 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+
+There are several distinct situations that can happen and each requires a bit different handling:
+
+1. The underlying library used by your extension is using MP Metrics API directly.
```
```suggestion
1. The underlying library used by your extension is using the MicroProfile Metrics API directly.
```
(for the first occurrence, this might not be obvious to the reader)

gsmet commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1290,6 +1290,143 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+but it is still possible to avoid the need for `smallrye-metrics` (which contains the implementation class of the metric registry) by introducing a custom "no-op"
+implementation of the `MetricRegistry` which can, for example, return `null` from all its methods. This no-op implementation should be added to the application's classes
+so that it can be injected instead of the regular `io.smallrye.metrics.MetricRegistryImpl`. An example how this was implemented can be found
+by looking at the SmallRye Fault Tolerance library and `smallrye-fault-tolerance` Quarkus extension.
```
Maybe point to the appropriate class in the GitHub repo?

I know it can be broken in the future but it seems a bit too hard to find it so I would do it anyway.

jmartisk replied:
Ok, I added a link to the `NoopMetricRegistry` class

gsmet commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1290,6 +1290,143 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+==== Case 2: The library provides its own metric API
+In this case, the extension can make use of the `quarkus-smallrye-metrics-spi` module to expose the metrics from the library using MP Metrics.
+
+The extension author should do the following:
```
```suggestion
You should do the following:
```
(you're using `your` everywhere after that). Another option would be to use `Extension authors` and move to `they/their/them`. Not sure how it is in the rest of the document.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -311,4 +325,200 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+            BuildProducer<MetricBuildItem> metrics) {
+        Metadata activeCountMetadata = Metadata.builder()
+                .withName("agroal.active.count")
+                .withDescription("Number active of connections. This connections are in use and not available to be acquired.")
```
```suggestion
                .withDescription("Number of active connections. These connections are in use and not available to be acquired.")
```

jmartisk replied:
Good catch, I admit I was mostly just copying javadoc from Hibernate codebase without thinking too hard about it. There are some typos apparently.

gsmet commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java:
```diff
@@ -0,0 +1,98 @@
+            case "reapCount":
+                return metrics.reapCount();
+            default:
+                throw new IllegalArgumentException("Unknown data source metric");
```
```suggestion
                throw new IllegalArgumentException("Unknown datasource metric");
```
(I know we're not very consistent with this especially with the class names being `DataSource` (this comes from javax.) but let's stick to `datasource` in text)

gsmet commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceBuildTimeConfig.java:
```diff
@@ -22,4 +22,11 @@
+     * smallrye-metrics extension is active.
+     */
+    @ConfigItem
+    public Optional<Boolean> enableMetrics;
```
So we don't have a way to disable metrics at runtime?

It seems like a very common request.

jmartisk replied:
Agroal seems to support disabling metrics dynamically, but I made this a build time property after discussing with Emmanuel who thinks we don't need that: https://github.com/quarkusio/quarkus/pull/5272#issuecomment-555106563

geoand replied:
I don't know if we should actually be able to turn it off at runtime given everything the user has to do to actually turn it on :). I mean it seems like the user really wants it explicitly. 

Now I am sure there scenaria were someone will want to turn it off, but perhaps we can keep it as is for now and reexamine if we have such a request in the future?

jmartisk replied:
Yeah I guess, I don't see much usefulness in dynamically turning off something that you had to explicitly turn on in the first place. 
And especially since turning off metric collection internally will mean that the already collected exposed metrics will suddenly "freeze" at their current values, which can become a pitfall, unless we add a hook that will also de-register the externally exposed Metrics as soon as this gets turned off.

gsmet replied:
So FTR, my use case was: metrics are not free, if you're in a constrained environment, you might want to disable them most of the time and enable them in case something is going wrong.

But yes, it can probably wait.

gsmet reviewed (changes requested):
I added a couple of suggestions and (I think) an important question.

jmartisk commented:
Hey @gsmet I've addressed your comments, does it look better to you now, please?

geoand commented:
@jmartisk just to be on the safe side, can you please rebase onto the latest master?

geoand commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1290,6 +1290,144 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+}
+----
+
+In this case, be aware that the MP Metrics API and implementation might not be available in the runtime classpath!
```
```suggestion
In this case, be aware that the MP Metrics API and implementation might not be available on the runtime classpath!
```

geoand commented on docs/src/main/asciidoc/writing-extensions.adoc:
```diff
@@ -1290,6 +1290,144 @@ HealthBuildItem addHealthCheck(AgroalBuildTimeConfig agroalBuildTimeConfig) {
+- The Metrics capability is present and therefore the SmallRye Metrics library is available
+- Metrics are not disabled for this extension
+
+A recommended way to implement this check would be to produce a specific `SystemPropertyBuildItem` during build, which will make it is easy
```
```suggestion
The recommended way to implement this check would be to produce a specific `SystemPropertyBuildItem` during build, which will make it easy
```

geoand commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -265,6 +265,24 @@ If you have multiple datasources, all datasources will be checked and the status
+collecting the metrics as well as exposing them in the `/metrics` endpoint, because it does not make sense to
+expose metrics if the mechanism to collect them is disabled.
+
+Conversely, setting `quarkus.datasource.enable-metrics` to `true` (or `quarkus.datasource.<datasource name>.enable-metrics` for a named datasource) explicitly can be used to enable collecting of metrics even if
```
```suggestion
Conversely, setting `quarkus.datasource.enable-metrics` to `true` (or `quarkus.datasource.<datasource name>.enable-metrics` for a named datasource) explicitly can be used to enable collection of metrics even if
```

geoand reviewed (commented):
I added some suggestions on the docs side of things. I'll also check out the code but I am pretty sure that after all the discussions things will be in place :)

geoand commented:
I also added the `breaking-change` label since the metrics property for the datasource config has moved to build time

geoand reviewed (approved):
I'm good with the state of this, thanks @jmartisk.

We probably want to squash the commits however

jmartisk commented:
Just working on squashing and rebasing, give me 3 minutes :)

jmartisk commented:
Done, I addressed @geoand 's comments, squashed it back into 3 commits and rebased against current master

geoand commented:
Excellent, thanks!

gsmet commented:
@geoand I approved it too. I let you merge it?

geoand commented:
Sure thing @gsmet!
